### PR TITLE
Fix mutex contention

### LINKED ISSUE #2751: Contention in futures aware mutex
it is possible to trigger cases where traffic will be highly synchronized, which means queries will reach the same places in code at the same time. When that happens, we can see high contention on parts guarded by futures aware mutexes, like the query planning deduplication.
We moved to these types of mutex, instead of `std` or `parking_lot` mutexes, due to the high risk of deadlocks across execution threads. Now that the code has settled down a bit, we can revisit them and move back to more performant solutions.

In general, the rule of thumb we should apply:

When the guarded section includes asynchronous code, use a futures aware mutex. If possible they should be avoided though.

```
...code
guard = mutex.lock().await

...code that awaits on something else

drop(guard)
```

When the guarded section only contains synchronous code, use a std mutex. Isolate the locking and unlocking in a non async function to make sure this is safe (anyway the compiler will complain if there's any risk this can happen because the mutex guard is not `Send`):

```
async fn do() {
  ...code
  sync_work()
  ...code that awaits on something else
}

fn sync_work() {
  guard = mutex.lock();
  ...work
  drop(guard)
}
```

If possible, we should make sure that the guarded section is short and small, like updating a value in a map, and does not include long running code sections.

### PR DESCRIPTION
**Pull request by Geal**
Fix #2751

We use futures aware mutexes in various places, but they are expensive and in some tests show high contentions. This is a serie of fixes, in deduplication and other parts of the code, to use std mutexes instead.

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

**Review by bnjjj — APPROVED** [review]

File: apollo-router/src/cache/mod.rs [hunk 1]
```diff
@@ -112,20 +101,60 @@ where
...
+                Ok(w)
+            }
+        }
+    }
+
+    fn remove_from_wait_map(wait_map: &WaitMap<K, V>, key: &K) {
+        let mut locked_wait_map = wait_map
+            .lock()
+            .expect("only locked in get_or_insert_wait_map or remove_from_wait_map");
+        let _ = locked_wait_map.remove(&key);
```
Comment (Geal) [inline apollo-router/src/cache/mod.rs:140]:
```suggestion
        let _ = locked_wait_map.remove(key);
```

**Comment by Geal** [comment]
sure, it's not urgent

**Comment by BrynCooke** [comment]
Perf test please

**Comment by BrynCooke** [comment]
Perf test please

**Comment by BrynCooke** [comment]
Perf test please

**Comment by BrynCooke** [comment]
I did quite a few test runs now, and it seems like a slight perf regression.
Eyeballing things it looks like max RPS is down about 150, also the behaviour when max is reached is more cliff like than on dev.

Suggest we break this up into smaller PRs and asses individually.

**Comment by Geal** [comment]
closing this as this is split in multiple independent PRs

**Pull request closed without merging**